Ambigous overload resolution

The following snippet fails to compile

```c++
#include "xtensor-python/pyarray.hpp"
#include "xtensor-python/pytensor.hpp"

auto xtensor_test(xt::pytensor<double, 3>& arr, xt::xtensor<double, 3> const& x) {
  arr /= x;
  return arr;
}
```

while this one is OK

```c++
#include "xtensor-python/pyarray.hpp"
#include "xtensor-python/pytensor.hpp"

auto xtensor_test(xt::pytensor<double, 3>& arr, xt::xtensor<double, 3> const& x) {
  arr = arr / x;
  return arr;
}
```

tested with clang 6 and gcc 8, xtensor 0.20.8 and xtensor-python 0.23.1

Error log:

```
In file included from /home/sguelton/.venvs/numpy-benchmarks/include/xtensor-python/pyarray.hpp:20,
                 from a.cpp:2:
/home/sguelton/.venvs/numpy-benchmarks/include/xtensor-python/pycontainer.hpp: In instantiation of ‘class xt::pycontainer<xt::pytensor<double, 3> >’:
/home/sguelton/.venvs/numpy-benchmarks/include/xtensor-python/pytensor.hpp:143:11:   required from ‘class xt::pytensor<double, 3>’
a.cpp:6:11:   required from here
/home/sguelton/.venvs/numpy-benchmarks/include/xtensor-python/pycontainer.hpp:53:11: warning: ‘xt::pycontainer<xt::pytensor<double, 3> >’ declared with greater visibility than its base ‘pybind11::object’ [-Wattributes]
     class pycontainer : public pybind11::object,
           ^~~~~~~~~~~
a.cpp: In function ‘auto xtensor_test(xt::pytensor<double, 3>&, xt::xtensor<double, 3>&)’:
a.cpp:6:11: error: request for member ‘operator/=’ is ambiguous
   arr /=  x;
           ^
In file included from /home/sguelton/.venvs/numpy-benchmarks/include/xtensor-python/pyarray.hpp:18,
                 from a.cpp:2:
/home/sguelton/.venvs/numpy-benchmarks/include/xtensor/xsemantic.hpp:74:23: note: candidates are: ‘template<class E> xt::xsemantic_base<D>::derived_type& xt::xsemantic_base<D>::operator/=(const xt::xexpression<E1>&) [with E = E; D = xt::pytensor<double, 3>]’
         derived_type& operator/=(const xexpression<E>&);
                       ^~~~~~~~
/home/sguelton/.venvs/numpy-benchmarks/include/xtensor/xsemantic.hpp:50:47: note:                 ‘template<class E> xt::disable_xexpression<E, typename xt::xexpression<E>::derived_type&> xt::xsemantic_base<D>::operator/=(const E&) [with E = E; D = xt::pytensor<double, 3>]’
         disable_xexpression<E, derived_type&> operator/=(const E&);
                                               ^~~~~~~~
In file included from /home/sguelton/.venvs/numpy-benchmarks/include/pybind11/cast.h:13,
                 from /home/sguelton/.venvs/numpy-benchmarks/include/pybind11/attr.h:13,
                 from /home/sguelton/.venvs/numpy-benchmarks/include/pybind11/pybind11.h:44,
                 from /home/sguelton/.venvs/numpy-benchmarks/include/pybind11/complex.h:12,
                 from /home/sguelton/.venvs/numpy-benchmarks/include/xtensor-python/pycontainer.hpp:16,
                 from /home/sguelton/.venvs/numpy-benchmarks/include/xtensor-python/pyarray.hpp:20,
                 from a.cpp:2:
/home/sguelton/.venvs/numpy-benchmarks/include/pybind11/pytypes.h:134:12: note:                 ‘pybind11::object pybind11::detail::object_api<Derived>::operator/=(const pybind11::detail::object_api<Derived>&) const [with Derived = pybind11::handle]’
     object operator/=(object_api const &other) const;
            ^~~~~~~~
```

--- a/.appveyor.yml
+++ b/.appveyor.yml
@@ -22,7 +22,7 @@ install:
   - conda update -q conda
   - conda info -a
   - conda install gtest cmake -c conda-forge
-  - conda install pytest numpy pybind11==2.2.4 xtensor==0.20.8 -c conda-forge
+  - conda install pytest numpy pybind11==2.4.2 xtensor==0.20.8 -c conda-forge
   - "set PYTHONHOME=%MINICONDA%"
   - cmake -G "NMake Makefiles" -D CMAKE_INSTALL_PREFIX=%MINICONDA%\\Library -D BUILD_TESTS=ON -D PYTHON_EXECUTABLE=%MINICONDA%\\python.exe .
   - nmake test_xtensor_python

--- a/include/xtensor-python/pyarray.hpp
+++ b/include/xtensor-python/pyarray.hpp
@@ -390,6 +390,20 @@ namespace xt
         static self_type ensure(pybind11::handle h);
         static bool check_(pybind11::handle h);
 
+#if PYBIND11_VERSION_MAJOR == 2 && PYBIND11_VERSION_MINOR >= 3
+        // Prevent ambiguous overload resolution for operators defined for
+        // both xt::xcontainer_semantic and pybind11::object.
+        using semantic_base::operator+=;
+        using semantic_base::operator-=;
+        using semantic_base::operator*=;
+        using semantic_base::operator/=;
+        using semantic_base::operator|=;
+        using semantic_base::operator&=;
+        using semantic_base::operator^=;
+        // using semantic_base::operator<<=;
+        // using semantic_base::operator>>=;
+#endif
+
     private:
 
         inner_shape_type m_shape;

--- a/include/xtensor-python/pycontainer.hpp
+++ b/include/xtensor-python/pycontainer.hpp
@@ -140,18 +140,6 @@ namespace xt
         using pybind11::object::operator^;
         using pybind11::object::operator<<;
         using pybind11::object::operator>>;
-
-        // Prevent ambiguous overload resolution for operators defined for
-        // both xt::xcontainer_semantic and pybind11::object.
-        using pybind11::object::operator+=;
-        using pybind11::object::operator-=;
-        using pybind11::object::operator*=;
-        using pybind11::object::operator/=;
-        using pybind11::object::operator|=;
-        using pybind11::object::operator&=;
-        using pybind11::object::operator^=;
-        using pybind11::object::operator<<=;
-        using pybind11::object::operator>>=;
 #endif
     };
 

--- a/include/xtensor-python/pycontainer.hpp
+++ b/include/xtensor-python/pycontainer.hpp
@@ -124,6 +124,35 @@ namespace xt
 
         PyArrayObject* python_array() const;
         size_type get_buffer_size() const;
+
+    private:
+
+#if PYBIND11_VERSION_MAJOR == 2 && PYBIND11_VERSION_MINOR >= 3
+        // Prevent ambiguous overload resolution for operators defined for
+        // both xt::xcontainer and pybind11::object.
+        using pybind11::object::operator~;
+        using pybind11::object::operator+;
+        using pybind11::object::operator-;
+        using pybind11::object::operator*;
+        using pybind11::object::operator/;
+        using pybind11::object::operator|;
+        using pybind11::object::operator&;
+        using pybind11::object::operator^;
+        using pybind11::object::operator<<;
+        using pybind11::object::operator>>;
+
+        // Prevent ambiguous overload resolution for operators defined for
+        // both xt::xcontainer_semantic and pybind11::object.
+        using pybind11::object::operator+=;
+        using pybind11::object::operator-=;
+        using pybind11::object::operator*=;
+        using pybind11::object::operator/=;
+        using pybind11::object::operator|=;
+        using pybind11::object::operator&=;
+        using pybind11::object::operator^=;
+        using pybind11::object::operator<<=;
+        using pybind11::object::operator>>=;
+#endif
     };
 
     namespace detail

--- a/include/xtensor-python/pytensor.hpp
+++ b/include/xtensor-python/pytensor.hpp
@@ -194,6 +194,20 @@ namespace xt
         static self_type ensure(pybind11::handle h);
         static bool check_(pybind11::handle h);
 
+#if PYBIND11_VERSION_MAJOR == 2 && PYBIND11_VERSION_MINOR >= 3
+        // Prevent ambiguous overload resolution for operators defined for
+        // both xt::xcontainer_semantic and pybind11::object.
+        using semantic_base::operator+=;
+        using semantic_base::operator-=;
+        using semantic_base::operator*=;
+        using semantic_base::operator/=;
+        using semantic_base::operator|=;
+        using semantic_base::operator&=;
+        using semantic_base::operator^=;
+        // using semantic_base::operator<<=;
+        // using semantic_base::operator>>=;
+#endif
+
     private:
 
         inner_shape_type m_shape;

--- a/test/test_pytensor.cpp
+++ b/test/test_pytensor.cpp
@@ -240,4 +240,26 @@ namespace xt
         auto v = xt::view(arr, xt::all());
         EXPECT_EQ(v(0), 0.);
     }
+
+    TEST(pytensor, unary)
+    {
+        pytensor<int, 1> a = { 1, 2, 3 };
+        pytensor<int, 1> res = -a;
+        pytensor<int, 1> ref = { -1, -2, -3 };
+        EXPECT_EQ(ref(0), res(0));
+        EXPECT_EQ(ref(1), res(1));
+        EXPECT_EQ(ref(1), res(1));
+    }
+
+    TEST(pytensor, inplace_pybind11_overload)
+    {
+        // pybind11 overrrides a number of operators in pybind11::object.
+        // This is testing that the right overload is picked up.
+        pytensor<double, 1> a = { 1.0, 2.0, 3.0 };
+        a /= 2;
+        pytensor<double, 1> ref = { 0.5, 1.0, 1.5 };
+        EXPECT_EQ(ref(0), a(0));
+        EXPECT_EQ(ref(1), a(1));
+        EXPECT_EQ(ref(1), a(1));
+    }
 }

